File: src/main/java/testbed/demo/tests/Raycast.java
package testbed.demo.tests;

import library.collision.Arbiter;
import library.dynamics.Body;
import library.rays.Ray;
import library.dynamics.World;
import library.math.Mat2;
import library.math.Vec2;
import testbed.demo.TestBedWindow;

public class Raycast {
    public static final String[] text = {"Raycast:"};
    public static boolean active = false;

    public static void load(TestBedWindow testBedWindow) {
        testBedWindow.setWorld(new World(new Vec2(0, -9.81)));
        testBedWindow.setCamera(new Vec2(-100, -20), 3.3);
        active = true;

        boolean isValid = false;
        while (!isValid) {
            isValid = true;
            testBedWindow.generateBoxOfObjects();
            for (Body b : testBedWindow.getWorld().getBodies()) {
                if (Arbiter.isPointInside(b, new Vec2())) {
                    isValid = false;
                    testBedWindow.getWorld().clearWorld();
                    break;
                }
            }
        }

        Ray r = new Ray(new Vec2(), new Vec2(0, 1), 1000);
        testBedWindow.add(r);
    }

    public static void action(Ray r) {
        Mat2 u = new Mat2();
        u.set(-0.0006);
        u.mul(r.getDirection());
    }
}

File: src/test/java/demo/tests/Raycast.kt
package demo.tests

import demo.window.TestBedWindow
import library.collision.Arbiter.Companion.isPointInside
import library.dynamics.World
import library.math.Mat2
import library.math.Vec2
import library.rays.Ray

object Raycast {
    @JvmField
    val text = arrayOf("Raycast:")
    @JvmField
    var active = false
    @JvmStatic
    fun load(testBedWindow: TestBedWindow) {
        testBedWindow.world = World(Vec2(.0, -9.81))
        testBedWindow.setCamera(Vec2(-100.0, -20.0), 3.3)
        active = true
        var isValid = false
        while (!isValid) {
            isValid = true
            testBedWindow.generateBoxOfObjects()
            for (b in testBedWindow.world.bodies) {
                if (isPointInside(b, Vec2())) {
                    isValid = false
                    testBedWindow.world.clearWorld()
                    break
                }
            }
        }
        val r = Ray(Vec2(), Vec2(.0, 1.0), 1000)
        testBedWindow.add(r)
    }

    @JvmStatic
    fun action(r: Ray) {
        val u = Mat2()
        u.set(-0.0006)
        u.mul(r.direction)
    }
}